File: cpp/54-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ret;
        int from_row = 0, from_col = 0;
        int to_row = matrix.size();
        int to_col = matrix[0].size();
        int size = to_row * to_col;
        int movr[4] = {0, 1, 0, -1};
        int movc[4] = {1, 0, -1, 0};
        int r = 0, c = 0, cr = 0, cc = 0, m = 0;
        
        for(int i = 0; i < size; i++){
            ret.push_back(matrix[r][c]);
            cr = r + movr[m];
            cc = c + movc[m];
            if(from_row <= cr && cr < to_row && from_col <= cc && cc < to_col){
                r = cr;
                c = cc;
            }
            else{
                if(m == 0)
                    from_row++;
                else if(m == 1)
                    to_col--;
                else if(m == 2)
                    to_row--;
                else if(m == 3)
                    from_col++;
                m = (m + 1) % 4;
                r = r + movr[m];
                c = c + movc[m];
            }
        }
        
        return ret;
    }
};

File: py/54-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        row = len(matrix)
        col = len(matrix[0])
        ret = []
        movr = [0, 1, 0, -1]
        movc = [1, 0, -1, 0]
        r = c = m = 0
        seen = [[False] * col for _ in range(row)]
        
        for _ in range(row * col):
            ret.append(matrix[r][c])
            seen[r][c] = True
            cr, cc = r + movr[m], c + movc[m]
            if 0 <= cr < row and 0 <= cc < col and not seen[cr][cc]:
                r, c = cr, cc
            else:
                m = (m + 1) % 4
                r, c = r + movr[m], c + movc[m]
        
        return ret